docs/pageserver-thread-mgmt.md

```diff
@@ -21,19 +26,80 @@ also a `shudown_watcher()` Future that can be used with `tokio::select!`
+drop the request-handling Future if the HTTP connection is lost. The
+HTTP handling code spawns a task for each request to shield the rest
+of the codebase from surprise cancellations.
+
+The downside of non-cancellation safe code is that you have to be very
+careful when using `tokio::select!` and other such functions that can
+cause Futures to be dropped. They can only be used with functions that
+are explicitly documented to be cancellation-safe, or you need to
+spawn a separate task to shield from the cancellation.
+
+#### How to cancel, then?
+
+If our code is not cancellation-safe, how do you cancel long-running
+tasks? Use CancellationTokens.
+
+TODO: More details on that. And we have an ongoing discussion on what
+to do if cancellations might come from multiple sources.
+
+
+#### Exceptions
+Some library functions are cancellation-safe, and are explicitly marked
+as such. For example, `utils::seqwait`.
+
+
+#### Rationale
+
+The alternative would be to make all async code cancellation-safe,
+unless otherwise marked. That way, you could use `tokio::select!` more
+liberally. The reasons we didn't choose that are explained in this
+section.
+
+Writing code in a cancellation-safe manner is tedious, as you need to
+scrutinize every `.await` and ensure that if the `.await` call never
+returns, the system is in a safe, consistent state. In some ways, you
+need to do that with `?` and early `returns`, too, but `.await`s are
+easier to miss. It is also easier to perform cleanup tasks when a
+function returns an `Err` than when an `.await` simply never
+returns. You can use `scopeguard` and Drop guards to perform cleanup
+tasks, but it is more tedious. An `.await` that never returns is more
+similar to a panic.
+
+Note that even if you only use building blocks that themselves are
+cancellation-safe, it doesn't mean that the code as whole is
+cancellation-safe. For example, consider the following code:
 
-We use async to wait for incoming data on network connections, and to
-perform other long-running operations. For example, each WAL receiver
-connection is handled by a tokio Task. Once a piece of WAL has been
-received from the network, the task calls the blocking functions in
-the Repository to process the WAL.
+```
+while let Some(i) = rx.recv().await {
+	if let Err(_) = tx.send(i).await {
+		println!("receiver dropped");
+		return;
+		}
+	}
+}
+```
 
-The core storage code in `layered_repository/` is synchronous, with
-blocking locks and I/O calls. The current model is that we consider
-disk I/Os to be short enough that we perform them while running in a
-Tokio task. If that becomes a problem, we should use `spawn_blocking`
-before entering the synchronous parts of the code, or switch to using
-tokio I/O functions.
+It reads messages from one channel, sends them to another channel. If
+this code is cancelled at the `tx.send(i).await`, the message read
+from the receiver is lost. That may or may not be OK, depending on the
+context.
```

Comment:
rx and tx close together makes it seem like these are for the same channel.

Let's call `rx` the `work_inbox` and `tx` the `results_outbox`.
To make it clear these are separate channels.